packages/wallet/src/unipig-wallet.ts

```diff
@@ -7,52 +7,191 @@ import {
+    const signedState: SignedStateReceipt = await this.rollupClient.getState(
+      account
+    )
+
+    await this.ovm.storeSignedStateReceipt(signedState)
+
+    // If valid, update known state
+    if (
+      (account in this.knownState &&
+        signedState.signature === EMPTY_AGGREGATOR_SIGNATURE) ||
+      (await this.ovm.isStateReceiptProvablyValid(
+        signedState.stateReceipt,
+        AGGREGATOR_ADDRESS
+      ))
+    ) {
+      this.knownState[account] = signedState.stateReceipt
+    }
+
+    return account in this.knownState ? this.knownState[account] : undefined
+  }
+
+  public async send(
+    tokenType: TokenType,
+    from: Address,
+    to: Address,
+    amount: number
+  ): Promise<void> {
+    const transaction: Transfer = {
+      sender: from,
+      recipient: to,
+      tokenType,
+      amount,
+    }
+    await this.submitTransaction(transaction)
+  }
+
+  public async swap(
+    tokenType: TokenType,
+    from: Address,
+    inputAmount: number,
+    minOutputAmount: number,
+    timeoutMillis: number
+  ): Promise<void> {
+    const transaction: Swap = {
+      sender: from,
+      tokenType,
+      inputAmount,
+      minOutputAmount,
+      timeout: timeoutMillis,
+    }
+    await this.submitTransaction(transaction)
+  }
+
+  public async requestFaucetFunds(
+    forAddress: Address,
+    amount: number = 10
+  ): Promise<void> {
+    const faucetRequest: FaucetRequest = {
+      sender: forAddress,
+      amount,
+    }
+    await this.submitTransaction(faucetRequest)
+  }
 
-    // TODO: We'll want to persist this State & Signature
+  private async submitTransaction(
+    transaction: RollupTransaction
+  ): Promise<void> {
+    let receipts: SignedStateReceipt[] = []
+    try {
+      if (isFaucetTransaction(transaction)) {
+        receipts = [
+          await this.rollupClient.requestFaucetFunds(transaction, this),
+        ]
+      } else {
+        receipts = await this.rollupClient.sendTransaction(transaction, this)
+      }
+    } catch (e) {
+      if (e instanceof SignatureError) {
+        logError(
+          log,
+          `Transaction was sent and a response was received, indicating it probably went through, but the signature on the response was invalid, so we cannot be sure. Tx: ${serializeObject(
+            transaction
+          )}`,
+          e
+        )
+      } else {
+        logError(
+          log,
+          `An error occurred sending transaction: Tx: ${serializeObject(
+            transaction
+          )}.`,
+          e
+        )
+      }
+      throw e
+    }
 
-    return !!state.stateReceipt.state
-      ? state.stateReceipt.state.balances
-      : undefined
+    try {
+      await Promise.all(
+        receipts.map((r) => this.ovm.storeSignedStateReceipt(r))
+      )
+    } catch (e) {
+      logError(
+        log,
+        `An error persisting transaction receipt for Tx: ${serializeObject(
+          transaction
+        )}.`,
+        e
+      )
+    }
```

Comment:
So I had some conversations with Ben about how to structure the wallet / client library. Would love to hear your input on this. That said, here's what we came up with:

1. The overall package which people install is `ovm` -- eg. `npm install ovm` (I'm in communication with npm about giving us control of that package name & will hopefully get it soon).
2. The ovm contains a keystore (aka wallet) where people hold their keys.
3. The ovm contains a "transitioner" 
4. The ovm contains a "state solver" -- this has all of the decider logic.

## ovm.transitioner
- send
- swap
- etc

This gets user intent and translates it into L2 actions (rollup being the way we translate for now). It also handles signing

## ovm.stateSolver
This contains 3 components:
1. Computability model
2. Game model
3. State enforcer 

The computability model & game model is what we've been building with the deciders. The 'computability model' is a model of what can and cannot be computed by other agents in our system (aka no one can compute our signature but us). The 'game model' is our model of rollup & of course in our case will be super simplistic.

The state enforcer is what sends fraud proofs to the mainchain. This also will be super simplistic but it can live in it's own object maybe to signify how important & distinct it's role is.

## It's a state transition system!
https://en.wikipedia.org/wiki/Transition_system

We've got state & transitions which are derived based on agent modeling / information hiding! Very cool!

## What this means
Even though we don't have "full" implementations of all components, it's important that we have the structure communicated so that folks can look at it and understand it's relation to our full vision.

The only changes here would be I would suggest renaming this class to `transitioner` (feel free to think of other names, but this one is nice because it preserves the concept of facilitating a "state transition", and then maybe renaming what is currently called `ovm` to instead `state solver` and putting the state enforcer inside that.

I've attached a photo of the high level picture & will put this into a better diagram soon.

Hugs!

![image](https://user-images.githubusercontent.com/706123/65349050-8bf00200-dbb0-11e9-90ed-2f76a052769f.png)